File: 2017-kotlin/src/main/java/Day4.kt
import java.util.stream.Collectors

class Day4: Day<String> {
    override fun parse(text: String): String {
        return text
    }

    override fun part1(input: String): Any {
        return input.split("\n").filter { !it.isEmpty() }.count { valid(it) }
    }

    private fun valid(input: String): Boolean {
        val words = input.split(" ").map { it.trim() }.filter { !it.isEmpty() }
        val nonDupWords = words.toSet()
        val valid = words.size == nonDupWords.size
        return valid
    }

    private fun valid2(input: String): Boolean {
        val words = input.split(" ").map { it.trim() }.filter { !it.isEmpty() }.map {
            it.chars().sorted().mapToObj { it.toString() }.collect(Collectors.joining())
        }
        val nonDupWords = words.toSet()
        val valid = words.size == nonDupWords.size
        return valid
    }

    override fun part2(input: String): Any {
        return input.split("\n").filter { !it.isEmpty() }.count { valid2(it) }
    }

}

File: 2019-typescript/src/Day4.ts
import Day from "./Day";

class Day4 implements Day<number[]> {
  parse(input: string): number[] {
    return input.split('-').map(s => parseInt(s, 10));
  }

  part1(input: number[]): any {
    return this.countMatching(input[0], input[1], (e: number[]) => true)
  }

  countMatching(lowest: number, highest: number, extraCondition: Function) {
    var count = 0;
    for (var a = 0; a <= 9; a++) {
      for (var b = 0; b <= 9; b++) {
        for (var c = 0; c <= 9; c++) {
          for (var d = 0; d <= 9; d++) {
            for (var e = 0; e <= 9; e++) {
              for (var f = 0; f <= 9; f++) {
                var digits = [a, b, c, d, e, f];
                let n = a*100000 + b*10000 + c*1000 + d*100 + e*10 + f;
                if (n >= lowest && n <= highest) {
                  let match = false, decreasing = false;
                  for (var i = 0; i < digits.length - 1; i++) {
                    if (digits[i] === digits[i + 1]) {
                      match = true;
                    }
                    if (digits[i + 1] < digits[i]) {
                      decreasing = true;
                      break;
                    }
                  }
                  if (match && !decreasing && extraCondition(digits)) {
                    count++;
                  }
                }
              }
            }
          }
        }
      }
    }
    return count;
  }

  part2(input: number[]): any {
    let isOK = (digits: number[]) => {
      var consecutiveNumber = digits[0];
      var consecutiveCount = 1;
      for (var i = 1; i < digits.length; i++) {
        if (digits[i] === consecutiveNumber) {
          consecutiveCount++;
        } else {
          if (consecutiveCount === 2) {
            return true;
          }
          consecutiveNumber = digits[i];
          consecutiveCount = 1;
        }
      }
      return consecutiveCount === 2;
    }
    return this.countMatching(input[0], input[1], isOK)
  }
}

export default Day4;
